### ISSUE TITLE
Loss names are not equal to the outputs dictionary keys of a model

### ISSUE THREAD

**Issue by szleb** [tensorflow/tensorflow#55435]
**System information**
- Have I written custom code: yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10.0.18363
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8.0
- Python version: 3.8.10
- CUDA/cuDNN version: none
- GPU model and memory: none

**Current behavior**

The names for the single losses of a multi output model correspond to the output 
layer names instead to the output names provided by the keys in the outputs dictionary.
I am not sure if this is the desired behavior.

**Expected behavior**

I would expect that the names of the losses are the same as the keys in the dictionary 
passed to the outputs argument when creating the model.

**Standalone code to reproduce the issue**

    import numpy as np
    from tensorflow.keras.layers import Dense, Input
    from tensorflow.keras.models import Model
    
    first_input = Input(1)
    x = Dense(100)(first_input)
    x = Dense(1)(x)
    
    second_input = Input(1)
    x2 = Dense(100)(second_input)
    x2 = Dense(1)(x2)
    
    model = Model(
        inputs={"first_input": first_input, "second_input": second_input},
        outputs={"first_output": x, "second_output": x2},
    )
    
    first_input_data = np.random.random(100)
    second_input_data = np.random.random(100)
    first_output_data = np.random.random(100)
    second_output_data = np.random.random(100)
    
    model.compile(loss="mse")
    
    model.fit(
        x={"first_input": first_input_data, "second_input": second_input_data},
        y={"first_output": first_output_data, "second_output": second_output_data},
        batch_size=2,
        epochs=2,
    )

**Other info / logs**

The loss names are visible in the following logging, when running the code example:

    Epoch 1/2
    50/50 [==============================] - 0s 673us/step - loss: 0.2440 - dense_1_loss: 0.1308 - dense_3_loss: 0.1131
    Epoch 2/2
    50/50 [==============================] - 0s 592us/step - loss: 0.1934 - dense_1_loss: 0.0973 - dense_3_loss: 0.0961

The problem is that I cannot see which loss belongs to which single output. 

Of course, my real use case is much more complicated, so that I cannot just name the output layers correctly.

**Comment by bhack** [comment]
Give a name to the output layers:
https://stackoverflow.com/questions/53504813/is-there-a-way-of-renaming-the-metrics-and-losses-of-a-keras-model

**Comment by sushreebarsa** [comment]
@szleb Could you please refer to the [comment](https://github.com/tensorflow/tensorflow/issues/55435#issuecomment-1083572061) above and let us know if it helps?
Thanks!

**Comment by szleb** [comment]
Here is another code example closer to my use case. I hope that makes things a bit clearer

    import numpy as np
    from tensorflow.keras.layers import Dense, Input, concatenate, multiply, subtract
    from tensorflow.keras.models import Model
    
    first_input = Input(1, name="first_input")
    x = Dense(100)(first_input)
    x = Dense(1, name="model_1_output")(x)
    
    model_1 = Model(
        name="model_1",
        inputs={"first_input": first_input},
        outputs={"model_1_output": x},
    )
    
    second_input = Input(1, name="second_input")
    third_input = Input(1, name="third_input")
    subtraction = subtract([second_input, third_input], name="subtraction")
    multiplication = multiply([second_input, third_input], name="multiplication")
    
    model_2 = Model(
        name="model_2",
        inputs={"second_input": second_input, "third_input": third_input},
        outputs={"subtraction": subtraction, "multiplication": multiplication},
    )
    
    input_1 = Input(1, name="input_1")
    input_2 = Input(1, name="input_2")
    layer_1 = model_1(input_1)
    layer_2 = model_2({"second_input": input_2, "third_input": layer_1["model_1_output"]})
    
    my_model = Model(
        name="my_model",
        inputs={"first_input": input_1, "second_input": input_2},
        outputs={
            "model_1_output": layer_1["model_1_output"],
            "multiplication": layer_2["multiplication"],
            "subtraction": layer_2["subtraction"],
        },
    )
    
    first_input_data = np.random.random(100)
    second_input_data = np.random.random(100)
    first_output_data = np.random.random(100)
    second_output_data = np.random.random(100)
    third_output_data = np.random.random(100)
    
    my_model.compile(loss="mse")
    
    my_model.fit(
        x={"first_input": first_input_data, "second_input": second_input_data},
        y={
            "model_1_output": first_output_data,
            "subtraction": third_output_data,
            "multiplication": second_output_data,
        },
        batch_size=2,
        epochs=2,
    )

This produces the following log:

    Epoch 1/2
    50/50 [==============================] - 0s 694us/step - loss: 0.7059 - model_1_loss: 0.1758 - model_2_loss: 0.2769 - model_2_1_loss: 0.2531
    Epoch 2/2
    50/50 [==============================] - 0s 653us/step - loss: 0.6777 - model_1_loss: 0.1460 - model_2_loss: 0.2485 - model_2_1_loss: 0.2832

Here I do not know any more which loss belongs to the multiplication layer and which to the subtraction layer.

I realised that in the code above `my_model.output_names =  ['model_1', 'model_2', 'model_2_1']`, but the keys of `my_model.output` and `my_model.output_shape` are the custom names `['model_1_output', 'multiplication', 'subtraction']`. I also figured out that when the model is created the outputs are ordered alphabetically with respect to the keys in the outputs dictionary (by tf.nest.flatten). So I am able to map the loss names to my output names, e.g. in a customized callback. In the given example this means that model_2_loss refers to the multiplication layer and model_2_1_loss to the subtraction layer. (Correct me if I am wrong.)

**Comment by bhack** [comment]
You need to use model outputs as a list not a dict:

https://github.com/keras-team/keras/blob/813e6911bd2edd78f85b0c245af6432bdd75cfc7/keras/engine/compile_utils.py#L659-L668

**Comment by szleb** [comment]
Seems not to solve the problem in this case:

    import numpy as np
    from tensorflow.keras.layers import Dense, Input, concatenate, multiply, subtract
    from tensorflow.keras.models import Model
    
    first_input = Input(1, name="first_input")
    x = Dense(100)(first_input)
    x = Dense(1, name="model_1_output")(x)
    
    model_1 = Model(
        name="model_1",
        inputs={"first_input": first_input},
        outputs=x,
    )
    
    second_input = Input(1, name="second_input")
    third_input = Input(1, name="third_input")
    subtraction = subtract([second_input, third_input], name="subtraction")
    multiplication = multiply([second_input, third_input], name="multiplication")
    
    model_2 = Model(
        name="model_2",
        inputs={"second_input": second_input, "third_input": third_input},
        outputs=[subtraction, multiplication],
    )
    
    input_1 = Input(1, name="input_1")
    input_2 = Input(1, name="input_2")
    layer_1 = model_1(input_1)
    layer_2 = model_2({"second_input": input_2, "third_input": layer_1})
    
    my_model = Model(
        name="my_model",
        inputs={"first_input": input_1, "second_input": input_2},
        outputs=[
            layer_1,
            layer_2[1],
            layer_2[0],
        ],
    )
    
    first_input_data = np.random.random(100)
    second_input_data = np.random.random(100)
    first_output_data = np.random.random(100)
    second_output_data = np.random.random(100)
    third_output_data = np.random.random(100)
    
    my_model.compile(loss="mse")
    
    my_model.fit(
        x={"first_input": first_input_data, "second_input": second_input_data},
        y=[
            first_output_data,
            third_output_data,
            second_output_data,
        ],
        batch_size=2,
        epochs=2,
    )

But still the ouptut looks like this:

	Epoch 1/2
	50/50 [==============================] - 0s 673us/step - loss: 0.5893 - model_1_loss: 0.1533 - model_2_loss: 0.1916 - model_2_1_loss: 0.2445
	Epoch 2/2
	50/50 [==============================] - 0s 612us/step - loss: 0.5682 - model_1_loss: 0.1377 - model_2_loss: 0.1807 - model_2_1_loss: 0.2498

The problem is that the inner model has two outputs and is used as a layer which covers the names of its output layers. Besides that I actually do use dictionaries on purpose no to depend on any order when defining the models and to get dictionaries when predicting data, so that the outputs can be identified.

**Comment by bhack** [comment]
I don't know if you can minimize your example a bit but the output names are

`print(my_model.output_names)`
`['model_1', 'model_2', 'model_2_1']`

**Comment by szleb** [comment]
Yes exactly. The ouptut_names are the names of the output layers and not the outputs-dictionary keys (when using dicts). Since the last layer is the model called `model_2` having two outputs, the output_names of the model are `model_2` and `model_2_1`.
Whatever the name of `model_2` is, I will never be able to identify the two outputs by their loss name. In the example with dictionary you also get

    print(my_model.output_names)
    ['model_1', 'model_2', 'model_2_1']

but

    print(my_model.output)
    {'model_1_output': <KerasTensor: shape=(None, 1) dtype=float32 (created by layer 'model_1')>, 
     'multiplication': <KerasTensor: shape=(None, 1) dtype=float32 (created by layer 'model_2')>, 
     'subtraction': <KerasTensor: shape=(None, 1) dtype=float32 (created by layer 'model_2')>}

Why it does not use the dicitionary keys for the `output_names` when provided? I think that is the point of the problem.

The long code example was just to show that it is not always possible to name the output layers correctly. The issue with the `output_names` can already be seen in the very first example of this issue.

**Comment by bhack** [comment]
Ok so this seems similari to https://github.com/tensorflow/tensorflow/issues/34114

**Comment by szleb** [comment]
changing the method _set_output_names in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/functional.py to something like the following would solve my problem:

    if type(self.output) is dict:
        self.output_names = list(self.output.keys())
        self.output_names.sort()
    else:
        <<current code>>

**Comment by bhack** [comment]
Keras now has its own repo so a ticket or PR need to be submited to Keras directly:
https://github.com/keras-team/keras

About this change I have some doubt related to the previous @qlzh727 evaluation:

https://github.com/tensorflow/tensorflow/issues/34114#issuecomment-588538295

**Comment by sachinprasadhs** [comment]
Thanks for opening this issue. Development of keras moved to separate repository https://github.com/keras-team/keras/issues

Please post this issue on keras-team/keras repo.
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!

**Comment by bhack** [comment]
https://github.com/keras-team/keras/issues/16405
